Dans ce TP, nous considérons des trajets en vélo partagé (similaire au vélib) en Californie. Deux jeux de données sont fournis : l'un qui contient les stations de vélo, l'autre, les trajets à vélo. Les déplacements à vélo se font d'une station à l'autre.

In [1]:
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .appName("Lab3TripApp") \
    .getOrCreate()

sc = spark.sparkContext

Charger le fichier station_data.csv dans un DataFrame station_df et le fichier trip_data.csv dans un DataFrame trip_df. Pour chaque Dataframe, il vous faudra demander une inférence des schémas et indiquer que la première ligne est un en-tête.

In [9]:
station_df = spark.read.csv("../../data/station_data.csv", header=True, inferSchema=True)
trip_df = spark.read.option("header", "true").option("inferSchema", "true").csv("../../data/trip_data.csv")

Afficher les schémas des 2 DataFrames.

In [10]:
station_df.printSchema()

root
 |-- station_id: integer (nullable = true)
 |-- name: string (nullable = true)
 |-- lat: double (nullable = true)
 |-- long: double (nullable = true)
 |-- dockcount: integer (nullable = true)
 |-- landmark: string (nullable = true)
 |-- installation: string (nullable = true)



In [11]:
station_df.toPandas()

station_id                               name        lat        long  \
0            2  San Jose Diridon Caltrain Station  37.329732 -121.901782   
1            3              San Jose Civic Center  37.330698 -121.888979   
2            4             Santa Clara at Almaden  37.333988 -121.894902   
3            5                   Adobe on Almaden  37.331415 -121.893200   
4            6                   San Pedro Square  37.336721 -121.894074   
..         ...                                ...        ...         ...   
65          77                  Market at Sansome  37.789625 -122.400811   
66          80    Santa Clara County Civic Center  37.352601 -121.905733   
67          82          Broadway St at Battery St  37.798541 -122.400862   
68          83                         Mezes Park  37.491269 -122.236234   
69          84                        Ryland Park  37.342725 -121.895617   

    dockcount       landmark installation  
0          27       San Jose     8/6/2013  
1          15       San Jose     8/5/2013  
2          11       San Jose     8/6/2013  
3          19       San Jose     8/5/2013  
4          15       San Jose     8/7/2013  
..        ...            ...          ...  
65         27  San Francisco    8/25/2013  
66         15       San Jose   12/31/2013  
67         15  San Francisco    1/22/2014  
68         15   Redwood City    2/20/2014  
69         15       San Jose     4/9/2014  

[70 rows x 7 columns]

In [15]:
station_df.describe().toPandas()

summary          station_id                                           name  \
0   count                  70                                             70   
1    mean                43.0                                           None   
2  stddev  24.166091947189145                                           None   
3     min                   2                                  2nd at Folsom   
4     max                  84  Yerba Buena Center of the Arts (3rd @ Howard)   

                   lat                 long          dockcount       landmark  \
0                   70                   70                 70             70   
1    37.59024338428572  -122.21841616428571  17.65714285714286           None   
2  0.20347253639672416   0.2094460497964454  4.010441857493954           None   
3            37.329732          -122.418954                 11  Mountain View   
4             37.80477          -121.877349                 27       San Jose   

  installation  
0           70  
1         None  
2         None  
3    1/22/2014  
4     8/7/2013

In [12]:
trip_df.printSchema()

root
 |-- TripID: integer (nullable = true)
 |-- Duration: integer (nullable = true)
 |-- StartDate: string (nullable = true)
 |-- StartStation: string (nullable = true)
 |-- StartTerminal: integer (nullable = true)
 |-- EndDate: string (nullable = true)
 |-- EndStation: string (nullable = true)
 |-- EndTerminal: integer (nullable = true)
 |-- Bike#: integer (nullable = true)
 |-- SubscriberType: string (nullable = true)
 |-- ZipCode: string (nullable = true)



In [13]:
trip_df.toPandas()

TripID  Duration        StartDate  \
0       913460       765  8/31/2015 23:26   
1       913459      1036  8/31/2015 23:11   
2       913455       307  8/31/2015 23:13   
3       913454       409  8/31/2015 23:10   
4       913453       789  8/31/2015 23:09   
...        ...       ...              ...   
354147  432951       619    9/1/2014 4:21   
354148  432950      6712    9/1/2014 3:16   
354149  432949       538    9/1/2014 0:05   
354150  432948       568    9/1/2014 0:05   
354151  432947       569    9/1/2014 0:05   

                                StartStation  StartTerminal          EndDate  \
0       Harry Bridges Plaza (Ferry Building)             50  8/31/2015 23:39   
1                San Antonio Shopping Center             31  8/31/2015 23:28   
2                             Post at Kearny             47  8/31/2015 23:18   
3                         San Jose City Hall             10  8/31/2015 23:17   
4                      Embarcadero at Folsom             51  8/31/2015 23:22   
...                                      ...            ...              ...   
354147                    Powell Street BART             39    9/1/2014 4:32   
354148  Harry Bridges Plaza (Ferry Building)             50    9/1/2014 5:08   
354149              South Van Ness at Market             66    9/1/2014 0:14   
354150              South Van Ness at Market             66    9/1/2014 0:15   
354151              South Van Ness at Market             66    9/1/2014 0:15   

                                      EndStation  EndTerminal  Bike#  \
0       San Francisco Caltrain (Townsend at 4th)           70    288   
1                        Mountain View City Hall           27     35   
2                              2nd at South Park           64    468   
3                            San Salvador at 1st            8     68   
4                         Embarcadero at Sansome           60    487   
...                                          ...          ...    ...   
354147                           Townsend at 7th           65    335   
354148  San Francisco Caltrain (Townsend at 4th)           70    259   
354149                             5th at Howard           57    466   
354150                             5th at Howard           57    461   
354151                             5th at Howard           57    318   

       SubscriberType ZipCode  
0          Subscriber    2139  
1          Subscriber   95032  
2          Subscriber   94107  
3          Subscriber   95113  
4            Customer    9069  
...               ...     ...  
354147     Subscriber   94118  
354148       Customer   44100  
354149       Customer      32  
354150       Customer      32  
354151       Customer      32  

[354152 rows x 11 columns]

In [16]:
trip_df.describe().toPandas()

summary             TripID            Duration      StartDate  \
0   count             354152              354152         354152   
1    mean    676962.27441607  1046.0326611172604           None   
2  stddev  138874.1535127174   30016.93615692944           None   
3     min             432947                  60  1/1/2015 0:25   
4     max             913460            17270400  9/9/2014 9:59   

                                    StartStation       StartTerminal  \
0                                         354152              354152   
1                                           None    58.4460175291965   
2                                           None  16.738860479692576   
3                                  2nd at Folsom                   2   
4  Yerba Buena Center of the Arts (3rd @ Howard)                  84   

         EndDate                                     EndStation  \
0         354152                                         354152   
1           None                                           None   
2           None                                           None   
3  1/1/2015 0:30                                  2nd at Folsom   
4  9/9/2014 9:59  Yerba Buena Center of the Arts (3rd @ Howard)   

          EndTerminal               Bike# SubscriberType               ZipCode  
0              354152              354152         354152                353874  
1   58.42181605638257   422.9579107275972           None     2613469.697174148  
2  16.876792344295154  159.84155003059414           None  1.4617651129228191E9  
3                   2                   9       Customer                     0  
4                  84                 878     Subscriber                   nil

Créer une vue pour chaque DataFrame.

In [17]:
station_df.createOrReplaceTempView("stations")
trip_df.createOrReplaceTempView("trips")

Trouver deux façons de calculer le nombre de trajets, l'une en appelant une méthode sur trip_df directement, l'autre en rédigeant une requête SQL de la vue correspondant au DataFrame tripData.

In [18]:
trip_df.count()

354152

In [20]:
spark.sql("SELECT count(*) FROM trips").show()

+--------+
|count(1)|
+--------+
|  354152|
+--------+



Ecrire une requête permettant de compter le nombre de trajets qui démarrent et se terminent à la même station.

In [21]:
spark.sql(
    """
    SELECT count(*) FROM trips
    WHERE StartTerminal=EndTerminal
    """
).show()

+--------+
|count(1)|
+--------+
|   10276|
+--------+



On souhaite désormais obtenir l’id des stations associées à ces trajets. Ecrire une requête renvoyant la liste des terminaux concernés ainsi que le nombre de trajets pour chacun de ces terminaux. Trier le résultat par ordre décroissant de nombre de trajets.
<br>Exemple de sortie :
<br>+--------+--------+
<br>|terminal|count(1)|
<br>+--------+--------+
<br>| 60| 850|
<br>| 50| 708|
<br>| 35| 348|
<br>| 76| 320|
<br>| 74| 307|
<br>(La station 60 est la plus concernée par ces trajets cycliques, avec 850 de ces trajets.)

In [ ]:
# Votre code ici

Dans la requête précédente, nous avons oublié un élément qui nous importe. Nous souhaitons compléter le résultat en indiquant le nombre de docks (dockcount) des stations concernées.
<br>Exemple de sortie :
<br>+--------+---------+--------+
<br>|terminal|dockcount|count(1)|
<br>+--------+---------+--------+
<br>| 60| 15| 850|
<br>| 50| 23| 708|
<br>| 35| 11| 348|
<br>| 76| 19| 320|
<br>| 74| 23| 307|
<br>Mettre à jour la requête.

Rédiger les 2 requêtes précédentes avec le DSL de DataFrame.

In [24]:
from pyspark.sql.functions import col, desc, count

In [26]:
result_df = trip_df.select(
    "StartTerminal"
).filter(
    col("StartTerminal") == col("EndTerminal")
).join(
    station_df, on=col("StartTerminal") == col("station_id")
).groupBy(
    "StartTerminal","dockcount"
).agg(
    count("*").alias("nombre")
).orderBy(
    desc("nombre")
)

In [27]:
result_df.show()

+-------------+---------+------+
|StartTerminal|dockcount|nombre|
+-------------+---------+------+
|           60|       15|   850|
|           50|       23|   708|
|           35|       11|   348|
|           76|       19|   320|
|           74|       23|   307|
|           39|       19|   296|
|           61|       27|   280|
|           67|       27|   277|
|           71|       19|   268|
|           70|       19|   260|
|           28|       23|   254|
|           48|       15|   248|
|           54|       15|   230|
|           69|       23|   227|
|           42|       15|   213|
|           73|       15|   200|
|           57|       15|   197|
|           64|       15|   194|
|            3|       15|   189|
|           72|       23|   181|
+-------------+---------+------+
only showing top 20 rows



Observer le plan d’exécution des requêtes.

In [29]:
result_df.explain(True)

== Parsed Logical Plan ==
'Sort ['nombre DESC NULLS LAST], true
+- Aggregate [StartTerminal#243, dockcount#212], [StartTerminal#243, dockcount#212, count(1) AS nombre#1820L]
   +- Join Inner, (StartTerminal#243 = station_id#208)
      :- Project [StartTerminal#243]
      :  +- Filter (StartTerminal#243 = EndTerminal#246)
      :     +- Project [StartTerminal#243, EndTerminal#246]
      :        +- Relation [TripID#239,Duration#240,StartDate#241,StartStation#242,StartTerminal#243,EndDate#244,EndStation#245,EndTerminal#246,Bike##247,SubscriberType#248,ZipCode#249] csv
      +- Relation [station_id#208,name#209,lat#210,long#211,dockcount#212,landmark#213,installation#214] csv

== Analyzed Logical Plan ==
StartTerminal: int, dockcount: int, nombre: bigint
Sort [nombre#1820L DESC NULLS LAST], true
+- Aggregate [StartTerminal#243, dockcount#212], [StartTerminal#243, dockcount#212, count(1) AS nombre#1820L]
   +- Join Inner, (StartTerminal#243 = station_id#208)
      :- Project [StartTerminal

Arrêter la session Spark.

In [ ]:
# Votre code ici